adds dependency analysis

juanvallejo opened the pull request:
Depends on https://github.com/openshift/origin/pull/18466
Depends on https://github.com/openshift/origin/pull/18606

Adds dependency graph analysis.
Outputs "yours", "mine", "ours" dependencies.

Usage:
```bash
$ ./depcheck analyze --root=github.com/openshift/origin --entry=cmd/... --entry=pkg/... --entry=tools/... --entry=test/... --openshift --dep=github.com/openshift/origin/vendor/k8s.io/kubernetes
```

[Output of the command above](https://gist.github.com/juanvallejo/d83d2f56e4f7345c3f37769237cb12a1)

cc @deads2k

deads2k commented:
graph still looks like you're cutting at github/org not github/org/repo

juanvallejo commented:
@deads2k updated graph to reflect latest changes in https://github.com/openshift/origin/pull/18466
Added "mine" and "ours" packages

deads2k commented:
remove all `gihtub.com/openshift/origin/...` packages from the yours/mine/ours output

deads2k commented on tools/depcheck/pkg/cmd/trace/pkg.go:
```diff
@@ -91,7 +91,10 @@ func BuildGraph(packages *PackageList, excludes []string) (*depgraph.MutableDire
 			to, exists := g.NodeByName(dependency)
 			if !exists {
-				return nil, fmt.Errorf("expected child node for dependency %q was not found in graph", dependency)
+				// if a package imports a dependency that we did not visit
```
this was supposed to end up in a separate pull. did I miss it?

juanvallejo replied:
Opened a pull here to begin addressing this https://github.com/openshift/origin/pull/18606

deads2k replied:
> Opened a pull here to begin addressing this #18606

whoa, that pull is big.  I was expecting it to be 5 lines long.

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,428 @@
+	return nil
+}
+
+func printYours(g *depgraph.MutableDirectedGraph, roots, nodes []graph.Node, you *depgraph.Node) ([]graph.Node, error) {
```
You have a print method that is mutating its input??  Don't do that.

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,428 @@
+		from := g.From(n)
+
+		// found an existing orphaned node to exclude
+		if len(to) == 0 && len(from) == 0 {
```
When does this happen?  I would expect these to snipped before you call this method.

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,428 @@
+	g.RemoveNode(you)
+
+	// find nodes not reachable from our given roots
+	yours := findOrphans(g, roots, nodes)
```
Given a clean starting graph, the idea of "roots" doesn't exist.

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,428 @@
+	for _, n := range nodes {
+		isOrphan := true
+		for _, root := range roots {
+			if hasPathFromTo(g, root, n) {
```
This doesn't look right. If you started your graph without orphans, then any orphans you're finding later on the result of the removing the "from" on an inbound link.

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,428 @@
+	return nil
+}
+
+func printYours(g *depgraph.MutableDirectedGraph, roots, nodes []graph.Node, you *depgraph.Node) ([]graph.Node, error) {
```
The logic in here looks wrong.  I expected you to produce a new graph by filtering out a selected set of nodes.  Then after you've built your new graph, you simply scan all nodes to see if they don't have any inbound links. You track those, then built a new graph using the nodes without any inbound links as your input.  Repeat until you have no changes.  Then union all the nodes you removed.

deads2k replied:
Notice that at no point do we discuss "roots".  I'm still trying to figure out why that concept is needed.

deads2k commented on tools/depcheck/pkg/graph/graph.go:
```diff
@@ -55,7 +55,64 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
 	return nil
 }
 
+func (g *MutableDirectedGraph) RemoveNode(n *Node) {
```
godoc whether this leaves dangling edges

deads2k commented on tools/depcheck/pkg/graph/graph.go:
```diff
@@ -55,7 +55,64 @@ func (g *MutableDirectedGraph) AddNode(n *Node) error {
 }
+
+// remove nodes with no inbound edges
+func (g *MutableDirectedGraph) PruneOrphans() {
```
Return a list of the nodes you removed.

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+// - "Yours": a list of every node in the set unique to the dependency tree of a given target node.
+// - "Mine": a list of every node in the set unique to the dependency tree of the root nodes (set non-overlapping with given target node)
+// - "Ours": a list of every node in the overlapping set between the dependency tree of the root nodes and a given target node
+func (o *TraceImportsOpts) analyzeGraph(g *depgraph.MutableDirectedGraph) error {
```
This is really hard to read.  Separate your library methods (things operating on input data), from your command wiring (things combining a set of library methods).  I should have a small method for identifying "which repos are only present because this set of nodes is present".

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+// This operation is done recursively until we no longer cause
+// any nodes to become orphaned after pruning the removed set.
+// The resulting set of total nodes recursively removed is returned.
+func calculateOrphans(g *depgraph.MutableDirectedGraph, targetNodes []*depgraph.Node) []*depgraph.Node {
```
I think this is the graph method I want.  Peer to the other graph methods I'd think.  Unit tests located there to check it.

deads2k replied:
Bad name.  This is `FindExclusiveDependencies(g, targetNodes)`

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+				continue
+			}
+
+			orphans = append(orphans, n.(*depgraph.Node))
```
Isn't this exactly what your PruneOrphans method does?  I didn't look, but for that method to work it would have to be recursive.  It is recursive, right?  If not, write the test that will cause that method to break and then fix it.

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+}
+
+type TraceImportsFlags struct {
+	Roots        []string
```
this should be dead, rigth?

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+	excludes []string
+	filters  []string
+
+	Analyze          string
```
bad name, what is this?

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+`
+
+type TraceImportsOpts struct {
+	Roots   []string
```
dead I hope

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+	fmt.Printf("Analyzing a total of %v packages\n", len(g.Nodes()))
+	fmt.Println()
+
+	yours := calculateOrphans(g, []*depgraph.Node{you})
```
Why are you messing with a single repo/node.  That doesn't make sense for what this needs to do.

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+	}
+
+	cmd.Flags().StringSliceVar(&flags.Roots, "root", flags.Roots, "set of entrypoints for dependency trees used to generate a depedency graph.")
+	cmd.Flags().StringVarP(&flags.Exclude, "exclude", "e", "", "json file containing a list of import-paths of packages to recursively exclude when traversing the set of given entrypoints specified through --root.")
```
description sounds bad.  I'd want to choose recursive exclusion or not based on `...` in golang, right?

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+	cmd.Flags().StringVarP(&flags.Exclude, "exclude", "e", "", "json file containing a list of import-paths of packages to recursively exclude when traversing the set of given entrypoints specified through --root.")
+	cmd.Flags().StringVarP(&flags.OutputFormat, "output", "o", "", "output generated dependency graph in specified format. One of: dot.")
+	cmd.Flags().StringVarP(&flags.Collapse, "collapse", "c", "", "json file containing a list of import-paths of packages to collapse sub-packages into.")
+	cmd.Flags().StringVarP(&flags.Analyze, "analyze", "a", "", "output a summary report on the dependency set of a given repository. Mutually exclusive with --output")
```
if its mutually exclusive, why do we have this lever?  Is this just a different output mode or is it (more likely) a different command.

deads2k commented on tools/depcheck/pkg/cmd/trace/trace.go:
```diff
@@ -0,0 +1,301 @@
+		},
+	}
+
+	cmd.Flags().StringSliceVar(&flags.Roots, "root", flags.Roots, "set of entrypoints for dependency trees used to generate a depedency graph.")
```
It really seems like you ought to have a set of common flags for building the graph, then a separate set of flags for analyzing it and that you're re-use the flags for building the graph in multiple different commands.

juanvallejo commented on tools/depcheck/pkg/cmd/analyze.go:
```diff
@@ -0,0 +1,273 @@
+
+	// TODO: populate mine with repoNodes, immediate vendor deps of repoNodes,
+	// and ensure no nodes overlap with those in the "yours" set.
+	mine := []*graph.Node{}
```
@deads2k I plan to address https://github.com/openshift/origin/pull/18466#issuecomment-368015835 here

juanvallejo commented on tools/depcheck/pkg/cmd/trace.go:
```diff
@@ -185,8 +185,31 @@ func (o *TraceOptions) Run() error {
+	// to have incomplete edge data based on the .go files
+	// immediately present in an entrypoint directory, and
+	// the direct import paths that they define.
+	possibleRoots := []string{}
```
@deads2k `pkg/oc` was being affected by this. 
The only parent `pkg/oc` has is `./cmd/oc`. 
`./cmd` was entrypoint, but was not tied to `cmd/oc` because there are no .go files under that directory that directly bring in `cmd/oc`. `cmd/oc` then had no "to" edges, and ended up being pruned - bringing `pkg/oc` with it.

juanvallejo commented:
@deads2k 
> I think the command is bugged somehow. I don't see a local user of github.com/boltdb/bolt. My guess is that the analysis is running transitives of the pkg, when we want to compare to directs.
>
> Build the "Mine" list by start with direct deps only.

Thanks, this is addressed in the latest commit in this pull. Here is the updated output from the command:
https://gist.github.com/juanvallejo/8d00fe04fdf742e35e4016dcf2d5c322

juanvallejo commented:
/retest

juanvallejo commented:
Updated graph
https://gist.github.com/juanvallejo/677c1c15d792717ba9c0767726d29162

deads2k commented:
> Updated graph
> https://gist.githubusercontent.com/juanvallejo/b45783567fac177554b27f2f5a4078bc/raw/af93842246abe2debbb222ba335151ddd3be0aae/graph.html (generated with d3.js, should be a bit more readable)

readable is clearly a relative statement

juanvallejo commented:
[Previous analysis output](https://gist.github.com/juanvallejo/8d00fe04fdf742e35e4016dcf2d5c322) generated with command:
```
$ ./depcheck analyze --root=github.com/openshift/origin --entry=cmd --entry=pkg  --exclude=./tools/depcheck/examples/origin_excludes.json --filter=./tools/depcheck/examples/origin_filters_extended.json --dep=github.com/openshift/origin/vendor/k8s.io/kubernetes
```

I have added the `test` and `tools` directories as additional entrypoints in the command below:
```
$ ./depcheck analyze --root=github.com/openshift/origin --entry=cmd --entry=pkg --entry=tools --entry=test --exclude=./tools/depcheck/examples/origin_excludes.json --filter=./tools/depcheck/examples/origin_filters_extended.json --dep=github.com/openshift/origin/vendor/k8s.io/kubernetes
```

Updated output: https://gist.github.com/juanvallejo/d83d2f56e4f7345c3f37769237cb12a1

juanvallejo commented:
@deads2k Went ahead and rebased with approved changes from https://github.com/openshift/origin/pull/18466.

Added the analysis bits on top.
Switched to using Dijkstra's instead of home-grown breath-first path traversal :) 
[Command output](https://gist.github.com/juanvallejo/d83d2f56e4f7345c3f37769237cb12a1) is exactly the same as the updated one posted in https://github.com/openshift/origin/pull/18543#issuecomment-369660707

juanvallejo commented:
/test extended_networking_minimal

deads2k commented:
prereq merged.  rebase for smaller diff

deads2k commented on tools/depcheck/pkg/cmd/analyze.go:
```diff
@@ -0,0 +1,211 @@
+`
+
+type AnalyzeOptions struct {
+	*graph.GraphOptions
```
no anonymous inclusion

deads2k commented on tools/depcheck/pkg/cmd/analyze.go:
```diff
@@ -0,0 +1,211 @@
+type AnalyzeOptions struct {
+	*graph.GraphOptions
+
+	// Packages to analyze against
```
against?

deads2k commented on tools/depcheck/pkg/cmd/analyze.go:
```diff
@@ -0,0 +1,211 @@
+}
+
+type AnalyzeFlags struct {
+	*graph.GraphFlags
```
no anonymous inclusion

deads2k commented on tools/depcheck/pkg/cmd/analyze.go:
```diff
@@ -0,0 +1,211 @@
+// - "Yours": a list of every node in the set unique to the dependency tree of a given target node.
+// - "Mine": a list of every node in the set unique to the dependency tree of the root nodes (set non-overlapping with given target node)
+// - "Ours": a list of every node in the overlapping set between the dependency tree of the root nodes and a given target node
+func (o *AnalyzeOptions) analyzeGraph(g *graph.MutableDirectedGraph) error {
```
needs unit tests

deads2k commented:
Add some tests.  Looks pretty good.

juanvallejo commented:
@deads2k thanks, added tests

juanvallejo commented:
@deads2k addressed in-person feedback regarding `calculateYours`, `calculateMineOurs` methods. Updated tests as well. Commits squashed.

juanvallejo commented:
/test unit
/test gcp

juanvallejo commented:
/test gcp

juanvallejo commented:
@deads2k tests passing, adding /lgtm
@soltysh fyi

/lgtm
